previous-version/previous_version.sh

```diff
@@ -0,0 +1,189 @@
+              IFS='.' read -r tag_major tag_minor tag_patch <<< "$tag"
+              IFS='.' read -r curr_major curr_minor curr_patch <<< "$current_tag"
+              curr_patch=${curr_patch%%-*}
+               if { ((tag_major == curr_major)) && ((tag_minor == curr_minor)) && ((tag_patch == curr_patch)); }; then
+                              continue
+                          fi
+
+              IFS='.' read -r candidate_major candidate_minor candidate_patch <<< "$candidate"
+
+              # Compare each previous tag with the current tag to find the closest preceding version
+              if ((tag_major == curr_major)); then
+                  if ((tag_minor == curr_minor)) && ((tag_patch <= curr_patch)) && ([[ -z "$candidate_patch" ]] || ((tag_patch > candidate_patch))); then
+                      candidate="$tag"
+                  elif ((tag_minor < curr_minor)) && ([[ -z "$candidate_minor" ]] || ((tag_minor > candidate_minor))); then
+                      candidate="$tag"
+                  fi
+              elif ((tag_major < curr_major)) && ([[ -z "$candidate_major" ]] || ((tag_major > candidate_major))); then
+                  candidate="$tag"
+              fi
+
+              if ((tag_patch == 0)); then
+                  IFS='.' read -r candidateWithPatch0_major candidateWithPatch0_minor _ <<< "$candidateWithPatch0"
+                  if ((tag_major == curr_major)) && ((tag_minor <= curr_minor)) && ([[ -z "$candidateWithPatch0_minor" ]] || ((tag_minor > candidateWithPatch0_minor))); then
+                      candidateWithPatch0="$tag"
+                  elif ((tag_major < curr_major)) && ([[ -z "$candidateWithPatch0_major" ]] || ((tag_major > candidateWithPatch0_major))); then
+                      candidateWithPatch0="$tag"
+                  fi
+              fi
+          fi
+      done
+
+      # Decide the previous tag based on whether the current patch number is 0
+      if [[ "$curr_patch" == "0" ]]; then
+          if [[ -n "$candidateWithPatch0" ]]; then
+              prev_tag="$candidateWithPatch0"
+          else
+              prev_tag="$candidate"
+          fi
+      else
+          prev_tag="$candidate"
+      fi
+       echo "$prev_tag"
+  }
+
+
+# Function to determine the previous tag based on the type and version
+determine_prev_tag() {
+  local current_tag="$1"
+  local type="$2"
+  local prev_tag=""
+  local -r prev_tags=$(git --no-pager tag --sort=-v:refname)
+
+  if [[ $type == "RC" ]]; then
+    # Extract the base version and RC version from the current tag
+    local base_version="${current_tag%-rc*}"
+    local -r current_rc_version=$(echo "$current_tag" | grep -oE "rc[0-9]+" | sed 's/rc//')
+
+    # Get the previous RC tag within the same minor version and with a lower RC number
+    prev_tag=$(get_previous_rc_tag "$base_version" "$current_rc_version" "$prev_tags")
+  fi
+
+  if [[ $type == "ALPHA" ]] || { [[ -z $prev_tag ]] && [[ $current_tag =~ -alpha[0-9]+(\.[0-9]+)?-rc[0-9]+$ ]]; }; then
```

Comment:
Also not sure about the syntax, what about:
```suggestion
  if [[ $type == "ALPHA" ]] || ( [[ -z $prev_tag ]] && [[ $current_tag =~ -alpha[0-9]+(\.[0-9]+)?-rc[0-9]+$ ]] ); then
```

Reply:
As far as I understand both curly braces and parentheses are used for command grouping, while the difference is that curly braces execute in same shell, while parentheses start a subshell. Not sure we need to start a subshell here

Reply:
Fair enough, let's keep it as is then.